WalletWasabi/WabiSabi/Client/AliceClient.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class AliceClient
+	{
+		public AliceClient(Guid aliceId, Guid roundId, ArenaClient arenaClient, IEnumerable<Coin> coins, FeeRate feeRate)
+		{
+			AliceId = aliceId;
+			RoundId = roundId;
+			ArenaClient = arenaClient;
+			Coins = coins;
+			FeeRate = feeRate;
+		}
+
+		public Guid AliceId { get; }
+		public Guid RoundId { get; }
+		private ArenaClient ArenaClient { get; }
+		private IEnumerable<Coin> Coins { get; }
+		private FeeRate FeeRate { get; }
+
+		public async Task ConfirmConnectionAsync(TimeSpan confirmInterval, CancellationToken token)
+		{
+			while (!await ConfirmConnectionAsync().ConfigureAwait(false))
+			{
+				await Task.Delay(confirmInterval, token).ConfigureAwait(false);
+			}
+		}
+
+		private async Task<bool> ConfirmConnectionAsync()
+		{
+			var inputWeight = 4 * Constants.P2wpkhInputVirtualSize;
+			var inputRemainingWeights = new[] { (long)ArenaClient.ProtocolMaxWeightPerAlice - inputWeight };
+
+			var amountCredentials = ArenaClient.AmountCredentialClient.Credentials;
+
+			var remainingFee = FeeRate.GetFee(Constants.P2wpkhInputVirtualSize);
+			List<Money> amounts = new();
+			foreach (Money amount in Coins.Select(c => c.Amount))
+			{
+				Money resultAmount = amount;
+				if (remainingFee > Money.Zero)
+				{
+					resultAmount = amount - remainingFee;
+					resultAmount = resultAmount < Money.Zero ? Money.Zero : resultAmount;
+
+					remainingFee -= amount;
+					remainingFee = remainingFee < Money.Zero ? Money.Zero : remainingFee;
+				}
+				amounts.Add(resultAmount);
+			}
```

Comment:
I would radically simplify this algorithm to request only one consolidated amount because otherwise there are many things that could be wrong, for example, we should remove amounts equal to zero, consolidate some amounts in case the number of amounts is greater than NumberOfCredentials and that could also introduce more issues:

Consider something like this:
  
```suggestion
			var totalFeeToPay = FeeRate.GetFee(Coins.Count() * Constants.P2wpkhInputVirtualSize);
			var totalAmount = Coins.Sum( coin => coin.Amount );
			var amountsToRequest = new[] { totalAmount - totalFeeToPay };
```

Reply:
agreed, for and for now it can just request the sum amount in the 1st credential and make the other credential requests 0, in the future `AliceClient` can be given the full list of registrations and the responsibility to calculate fees can be moved upwards to the per-round client if it makes sense.

Reply:
More general and addresses the previous suggestion as well about `coinCount`:

```cs
var totalInputVsize = Coins.Sum(x => x.EstimateInputVsize()); // see #5570 for why better than the constant

var totalInputFees = FeeRate.GetFee(totalInputVsize);

var totalAmount = Coins.Sum(x => x.Amount);

var inputRemainingWeights = new[] { (long)ArenaClient.ProtocolMaxWeightPerAlice - 4*totalInputVsize
 }; // or 4 * ((long)PerAliceVsizeAllocation - totalInputVsize ) after #5551
var amountsToRequest = new[] { totalAmount - totalInputFees };
```

Reply:
ACK